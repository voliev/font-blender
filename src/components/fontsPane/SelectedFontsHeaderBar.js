import React from "react";
import classNames from "classnames";
import ExpandLessRounded from "@material-ui/icons/ExpandLessRounded";
import ExpandMoreRounded from "@material-ui/icons/ExpandMoreRounded";
import PropTypes from "prop-types";
import "./styles/SelectedFontsHeaderBar.css";

const SelectedFontsHeaderBar = ({
  folded,
  selectedFonts,
  headerBarClickHandler
}) => {
  const componentClasses = classNames({
    SelectedFontsHeaderBar: true,
    empty: selectedFonts.length === 0
  });

  return (
    <div className={componentClasses} onClick={headerBarClickHandler}>
      <p className="text">
        <span className="num">{selectedFonts.length}</span>
        <span>
          {" "}
          {selectedFonts.length === 1 ? "family" : "families"} selected
        </span>
      </p>
      <div className="icon">
        {folded ? <ExpandLessRounded /> : <ExpandMoreRounded />}
      </div>
    </div>
  );
};

SelectedFontsHeaderBar.propTypes = {
  folded: PropTypes.bool.isRequired,
  selectedFonts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  headerBarClickHandler: PropTypes.func.isRequired
};

export default SelectedFontsHeaderBar;
