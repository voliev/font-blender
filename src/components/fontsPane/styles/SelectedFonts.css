.SelectedFonts {
  background-color: var(--white);
  grid-area: selected;
  left: 0;
  max-height: calc(100vh - 4rem);
  position: absolute;
  right: 0;
  /* top: 0 - Keeps the list hidden when first font families */
  /*          added after initial component render */
  top: 0;
  width: var(--side-pane-width);
  z-index: 100;
}

.SelectedFonts .main {
  display: flex;
  flex-direction: column;
  max-height: calc(100vh - 7rem);
}

/*
  Display viewport overlay when list
  of selected fonts is unfolded
 */
.SelectedFonts.empty~.overlay,
.SelectedFonts:not(.active)~.overlay,
.SelectedFonts:not(.unfolding)~.overlay,
.SelectedFonts:not(.folding)~.overlay,
.SelectedFonts:not(.folding-to-clear)~.overlay {
  background-color: none;
}

.SelectedFonts.active~.overlay,
.SelectedFonts.unfolding~.overlay,
.SelectedFonts.folding~.overlay,
.SelectedFonts.folding-to-clear~.overlay {
  background-color: var(--off-black);
  bottom: 0;
  height: 100%;
  left: 0;
  opacity: 0.6;
  position: fixed;
  right: 0;
  top: 0;
  transition: background-color 0.2s;
  width: 100%;
  z-index: 50;
}

.SelectedFonts.empty {
  bottom: auto;
  top: 0;
}

.SelectedFonts.folded {
  bottom: auto;
  top: 0;
  transition: initial;
  transform: none;
}

.SelectedFonts.unfolding {
  bottom: auto;
  top: 0;
  transition: transform 0.2s;
}

.SelectedFonts.active {
  bottom: 0;
  top: auto;
  transition: initial;
  transform: none;
}

.SelectedFonts.folding,
.SelectedFonts.folding-to-clear {
  bottom: 0;
  top: auto;
  transition: transform 0.2s;
}
