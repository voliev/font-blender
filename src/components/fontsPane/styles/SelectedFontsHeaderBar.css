.SelectedFontsHeaderBar {
  align-items: center;
  background: var(--blue);
  color: var(--white);
  cursor: pointer;
  display: flex;
  height: 3rem;
  justify-content: space-between;
  transition: background .3s, color .3s;
}

.SelectedFontsHeaderBar .text {
  align-items: center;
  display: flex;
  font-weight: 700;
  height: 100%;
  margin: 0 0 0 var(--space-lg);
}

.SelectedFontsHeaderBar .num {
  align-items: center;
  background: var(--white);
  border-radius: var(--border-radius-max);
  color: var(--blue);
  display: flex;
  height: 2rem;
  justify-content: center;
  margin-right: var(--space-xs);
  width: 2rem;
}

.SelectedFontsHeaderBar .icon {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: center;
  margin: 0 var(--space-lg) 0 0;
}

.SelectedFontsHeaderBar.empty {
  background: var(--dark-grey);
}

.SelectedFontsHeaderBar.empty .num {
  color: var(--dark-grey);
}
