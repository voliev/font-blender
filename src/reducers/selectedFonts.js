import undoable, { includeAction } from "redux-undo";
import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { getFonts } from "./fonts";

const initialState = [];

export const selectedFonts = (state = initialState, action) => {
  switch (action.type) {
    case ADD_SELECTED_FONT:
      return [...state, action.family];
    case REMOVE_FONT_FROM_SELECTED:
      return state.filter(family => family !== action.family);
    case CLEAR_SELECTED_FONTS:
      return [];
    default:
      return state;
  }
};

export default undoable(selectedFonts, {
  limit: 1,
  filter: includeAction([REMOVE_FONT_FROM_SELECTED, CLEAR_SELECTED_FONTS]),
  undoType: UNDO_DESELECTED
});

export const getSelectedFontFamilies = state => state.selectedFonts.present;
export const getSelectedFonts = state => {
  const fonts = getFonts(state);

  return getSelectedFontFamilies(state).reduce((selected, family) => {
    selected[family] = fonts[family];
    return selected;
  }, {});
};
export const isSelected = (state, family) =>
  state.selectedFonts.present.includes(family);
