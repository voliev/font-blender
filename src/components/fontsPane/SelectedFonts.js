import React, { Component, Fragment } from "react";
import classNames from "classnames";
import OutsideClickHandler from "react-outside-click-handler";
import PropTypes from "prop-types";
import SelectedFontsHeaderBar from "./SelectedFontsHeaderBar";
import SelectedFontsList from "./SelectedFontsList";
import SelectedFontsFooter from "./SelectedFontsFooter";
import "./styles/SelectedFonts.css";

class SelectedFonts extends Component {
  constructor() {
    super();

    this.state = {
      stage: 0
    };

    this.main = React.createRef();

    this.handleOutsideClick = this.handleOutsideClick.bind(this);
    this.handleHeaderBarClick = this.handleHeaderBarClick.bind(this);
    this.handleFontDeselection = this.handleFontDeselection.bind(this);
    this.handleClearSelectedFontsButtonClick = this.handleClearSelectedFontsButtonClick.bind(
      this
    );
    this.handleTransitionEnd = this.handleTransitionEnd.bind(this);
  }

  getLifecycleStage() {
    const stages = {
      0: "empty",
      1: "folded",
      2: "unfolding",
      3: "active",
      4: "folding",
      5: "folding-to-clear"
    };

    return stages[this.state.stage];
  }

  handleOutsideClick(e) {
    const { target } = e;
    const stage = this.getLifecycleStage();

    if (target.value !== "undo" && stage === "active") {
      this.setState({
        stage: 4
      });
    }
  }

  handleHeaderBarClick(e) {
    if (!e.target.closest(".SelectedFontsHeaderBar")) {
      return;
    }

    const stage = this.getLifecycleStage();

    if (stage === "folded") {
      this.setState({
        stage: 2
      });
    }

    if (stage === "active") {
      this.setState({
        stage: 4
      });
    }
  }

  handleFontDeselection(family) {
    const stage = this.getLifecycleStage();
    const { removeFontFromSelected, selectedFonts } = this.props;

    if (selectedFonts.length === 1 && stage === "active") {
      this.setState({
        stage: 5
      });
    }

    if (selectedFonts.length > 1 && stage === "active") {
      removeFontFromSelected(family);
    }
  }

  handleClearSelectedFontsButtonClick(e) {
    const { value } = e.target;
    const stage = this.getLifecycleStage();

    if (value === "clear" && stage === "active") {
      this.setState({
        stage: 5
      });
    }
  }

  handleTransitionEnd(e) {
    const { propertyName, target } = e;
    const classesArr = target.className.split(" ");
    const hasCorrespondingClass = classesArr.some(cn => cn === "SelectedFonts");

    if (propertyName !== "transform" && !hasCorrespondingClass) {
      return;
    }

    const stage = this.getLifecycleStage();
    const {
      clearSelectedFonts,
      removeFontFromSelected,
      selectedFonts
    } = this.props;

    if (stage === "unfolding") {
      this.setState({
        stage: 3
      });
    }

    if (stage === "folding") {
      this.setState({
        stage: 1
      });
    }

    if (stage === "folding-to-clear") {
      if (this.props.selectedFonts.length === 1) {
        removeFontFromSelected(selectedFonts[0]);
      } else {
        clearSelectedFonts();
      }

      this.setState({
        stage: 0
      });
    }
  }

  setComponentsCSSTransform() {
    const stage = this.getLifecycleStage();

    switch (stage) {
      case "unfolding":
        return {
          transform: `translateY(-${this.main.current.clientHeight}px)`
        };
      case "folding":
      case "folding-to-clear":
        return {
          transform: `translateY(${this.main.current.clientHeight}px)`
        };
      default:
        return {};
    }
  }

  componentDidUpdate({ selectedFonts: prevSelectedFonts }) {
    const { selectedFonts } = this.props;
    if (prevSelectedFonts.length === 0 && selectedFonts.length > 0) {
      this.setState({
        stage: 1
      });
    }

    if (prevSelectedFonts.length > 0 && selectedFonts.length === 0) {
      this.setState({
        stage: 0
      });
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { selectedFonts } = this.props;
    const { selectedFonts: nextSelectedFonts } = nextProps;

    return (
      selectedFonts.length !== nextSelectedFonts.length ||
      this.state.stage !== nextState.stage
    );
  }

  render() {
    const { stage } = this.state;
    const { selectedFonts } = this.props;

    const componentClasses = classNames({
      SelectedFonts: true,
      [this.getLifecycleStage()]: true
    });

    const componentStyles = this.setComponentsCSSTransform();

    return (
      <Fragment>
        <div
          className={componentClasses}
          onTransitionEnd={this.handleTransitionEnd}
          style={componentStyles}
          title={
            (stage === 0 && "Select fonts first") ||
            (stage === 1 && "Click to Expand") ||
            (stage === 3 && "Click to Fold") ||
            ""
          }
        >
          <OutsideClickHandler
            disabled={selectedFonts.length === 0 || stage === 0 || stage === 1}
            onOutsideClick={this.handleOutsideClick}
          >
            <SelectedFontsHeaderBar
              folded={stage < 3}
              headerBarClickHandler={this.handleHeaderBarClick}
              selectedFonts={selectedFonts}
            />
            {selectedFonts.length > 0 ? (
              <div className="main" ref={this.main}>
                <SelectedFontsList
                  fontDeselectionHandler={this.handleFontDeselection}
                  selectedFonts={selectedFonts}
                />
                <SelectedFontsFooter
                  clearSelectedFontsButtonClickHandler={
                    this.handleClearSelectedFontsButtonClick
                  }
                />
              </div>
            ) : null}
          </OutsideClickHandler>
        </div>
        <div className="overlay" />
      </Fragment>
    );
  }
}

SelectedFonts.propTypes = {
  selectedFonts: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  removeFontFromSelected: PropTypes.func.isRequired,
  clearSelectedFonts: PropTypes.func.isRequired
};

export default SelectedFonts;
