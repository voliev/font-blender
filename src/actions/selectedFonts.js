import {
  ADD_SELECTED_FONT,
  CLEAR_SELECTED_FONTS,
  REMOVE_FONT_FROM_SELECTED,
  UNDO_DESELECTED
} from "../constants";
import { startFetchFont } from "./fonts";
import { addNotification, removeNotification } from "./notifications";
import { getFonts } from "../reducers/fonts";

export const addFontToSelected = family => ({
  type: ADD_SELECTED_FONT,
  family
});

export const removeFontFromSelected = family => ({
  type: REMOVE_FONT_FROM_SELECTED,
  family
});

export const clearSelectedFonts = () => ({
  type: CLEAR_SELECTED_FONTS
});

export const startAddFontToSelected = family => {
  return (dispatch, getState) => {
    const state = getState();
    const fonts = getFonts(state);
    const font = fonts[family];

    if (font.fetched) {
      // Add font to selected if it has been already fetchedselected list
      dispatch(addFontToSelected(family));
    } else {
      // Fetch font first than add it to selected
      return dispatch(startFetchFont(family))
        .then(() => dispatch(addFontToSelected(family)))
        .catch(() => {});
    }
  };
};

export const undoDeselected = () => ({
  type: UNDO_DESELECTED
});

export const startRemoveFontFromSelected = family => {
  return dispatch => {
    dispatch(removeFontFromSelected(family));

    const text = `${family} font family has been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startClearSelectedFonts = () => {
  return (dispatch, getState) => {
    const state = getState();
    const selectedFonts = [...state.selectedFonts.present];

    dispatch(clearSelectedFonts());

    const text =
      selectedFonts.length > 1
        ? `${selectedFonts.length} font families have been removed`
        : `${selectedFonts.length} font family have been removed`;
    const group = "selectedFonts";
    // Add notification to display it with "undo" option
    dispatch(addNotification(text, group));
  };
};

export const startUndoDeselected = () => {
  return dispatch => {
    // Undo deselection
    dispatch(undoDeselected());

    const group = "selectedFonts";

    // Remove notification to hide "undo" option
    dispatch(removeNotification(group));
  };
};
