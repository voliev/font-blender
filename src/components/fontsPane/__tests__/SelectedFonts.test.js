import React from "react";
import { mount, shallow } from "enzyme";
import OutsideClickHandler from "react-outside-click-handler";
import SelectedFonts from "../SelectedFonts";
import SelectedFontsHeaderBar from "../SelectedFontsHeaderBar";
import selectedFonts from "../../../fixtures/selectedFonts";

const setup = (
  renderFn = shallow,
  propsOverrides = {},
  container = undefined
) => {
  const props = {
    selectedFonts: [],
    clearSelectedFonts: jest.fn(),
    removeFontFromSelected: jest.fn(),
    ...propsOverrides
  };

  return {
    ...props,
    wrapper: container
      ? renderFn(<SelectedFonts {...props} />, { attachTo: container })
      : renderFn(<SelectedFonts {...props} />)
  };
};

afterEach(() => {
  jest.clearAllMocks();
});

it("should render without crashing", () => {
  const { wrapper } = setup();

  expect(wrapper).toMatchSnapshot();
});

it("should not crash if the list is empty", () => {
  const { wrapper } = setup(shallow, { selectedFonts: [] });

  expect(wrapper).toMatchSnapshot();
});

it("should receive corresponding props", () => {
  const { clearSelectedFonts, removeFontFromSelected, wrapper } = setup(mount);

  expect(typeof wrapper.prop("selectedFonts")).toBe("object");
  expect(wrapper.prop("selectedFonts").hasOwnProperty("length")).toBeTruthy();

  expect(typeof wrapper.prop("clearSelectedFonts")).toBe("function");
  expect(wrapper.prop("clearSelectedFonts")).toBe(clearSelectedFonts);

  expect(typeof wrapper.prop("removeFontFromSelected")).toBe("function");
  expect(wrapper.prop("removeFontFromSelected")).toBe(removeFontFromSelected);
});

it("should initially render empty list", () => {
  const { wrapper } = setup();

  expect(wrapper.instance().getLifecycleStage()).toBe("empty");
  expect(wrapper.find("main")).toHaveLength(0);
});

it("should properly set OutsideClickHandler's 'disabled' value", () => {
  const { wrapper } = setup();

  expect(wrapper.find(OutsideClickHandler).prop("disabled")).toBeTruthy();

  wrapper.setProps({ selectedFonts });
  wrapper.setState({ stage: 3 });

  expect(wrapper.find(OutsideClickHandler).prop("disabled")).toBeFalsy();

  wrapper.setState({ stage: 1 });

  expect(wrapper.find(OutsideClickHandler).prop("disabled")).toBeTruthy();
});

it("should change component's lifecycle stage on header bar click", () => {
  const { wrapper } = setup(mount);

  expect(wrapper.state("stage")).toBe(0);

  wrapper.setProps({ selectedFonts });

  expect(wrapper.state("stage")).toBe(1);

  const event = {
    target: {
      closest: () => false
    }
  };

  wrapper.find(SelectedFontsHeaderBar).prop("headerBarClickHandler")(event);

  expect(wrapper.state("stage")).toBe(1);

  event.target.closest = () => true;
  wrapper.find(SelectedFontsHeaderBar).prop("headerBarClickHandler")(event);

  expect(wrapper.state("stage")).toBe(2);

  wrapper.setState({ stage: 3 });
  wrapper.find(SelectedFontsHeaderBar).prop("headerBarClickHandler")(event);

  expect(wrapper.state("stage")).toBe(4);
});

it("should return corresponding stage string value", () => {
  const getLifecycleStage = jest.spyOn(
    SelectedFonts.prototype,
    "getLifecycleStage"
  );
  const { wrapper } = setup(mount);
  const stages = [
    "empty",
    "folded",
    "unfolding",
    "active",
    "folding",
    "folding-to-clear"
  ];

  stages.forEach((stage, idx) => {
    if (wrapper.state("stage") === 1) {
      wrapper.setProps({ selectedFonts });
    }

    if (idx > 0) {
      wrapper.setState({ stage: idx });
    }
    wrapper.instance().getLifecycleStage();

    expect(getLifecycleStage).toHaveLastReturnedWith(stage);
  });
});

it("should handle font deselection", () => {
  const { wrapper, removeFontFromSelected } = setup(mount);
  const family = "Roboto";

  wrapper.setProps({ selectedFonts: [family] });

  expect(wrapper.instance().getLifecycleStage()).toBe("folded");

  wrapper.setState({ stage: 3 });

  wrapper.instance().handleFontDeselection(family);

  expect(wrapper.instance().getLifecycleStage()).toBe("folding-to-clear");

  wrapper.setProps({ selectedFonts });
  wrapper.setState({ stage: 3 });

  wrapper.instance().handleFontDeselection(family);

  expect(removeFontFromSelected).toHaveBeenCalledTimes(1);
  expect(removeFontFromSelected).toHaveBeenCalledWith(family);
});

it("should change state on clearing selected fonts' list", () => {
  const { wrapper } = setup(mount);

  wrapper.setProps({ selectedFonts });

  const event = {
    target: {}
  };

  wrapper.instance().handleClearSelectedFontsButtonClick(event);

  expect(wrapper.state("stage")).toBe(1);

  wrapper.setState({ stage: 3 });

  event.target = { value: "clear" };

  wrapper.instance().handleClearSelectedFontsButtonClick(event);

  expect(wrapper.state("stage")).toBe(5);
});

it("should handle list transition end event", () => {
  const { clearSelectedFonts, removeFontFromSelected, wrapper } = setup(mount);
  const event = {
    propertyName: "height",
    target: {
      className: "SelectedFontsList"
    }
  };

  wrapper.setProps({ selectedFonts });
  wrapper.setState({ stage: 2 });
  wrapper.instance().handleTransitionEnd(event);

  expect(wrapper.state("stage")).toBe(2);

  event.propertyName = "transform";
  event.target = {
    className: "SelectedFonts unfolding"
  };

  wrapper.instance().handleTransitionEnd(event);

  expect(wrapper.state("stage")).toBe(3);

  wrapper.setState({ stage: 4 });
  wrapper.instance().handleTransitionEnd(event);

  expect(wrapper.state("stage")).toBe(1);

  wrapper.setState({ stage: 5 });
  wrapper.instance().handleTransitionEnd(event);

  expect(clearSelectedFonts).toHaveBeenCalledTimes(1);
  expect(wrapper.state("stage")).toBe(0);

  wrapper.setState({ stage: 5 });
  wrapper.setProps({ selectedFonts: ["Roboto"] });
  wrapper.instance().handleTransitionEnd(event);

  expect(removeFontFromSelected).toHaveBeenCalledTimes(1);
  expect(removeFontFromSelected).toHaveBeenLastCalledWith("Roboto");
  expect(wrapper.state("stage")).toBe(0);
});

it("should handle click outside of the list", () => {
  const { wrapper } = setup(mount);
  const event = {
    target: {
      value: "undo"
    }
  };

  wrapper.setProps({ selectedFonts });
  wrapper.setState({ stage: 3 });
  wrapper.instance().handleOutsideClick(event);

  expect(wrapper.state("stage")).toBe(3);

  event.target = {};
  wrapper.instance().handleOutsideClick(event);

  expect(wrapper.state("stage")).toBe(4);
});

it("should set corresponding style for list animating", () => {
  const setComponentsCSSTransform = jest.spyOn(
    SelectedFonts.prototype,
    "setComponentsCSSTransform"
  );
  const jsdom = require("jsdom");
  const { JSDOM } = jsdom;
  const dom = new JSDOM("<div class='target'></div>");

  Object.defineProperty(dom.window.HTMLElement.prototype, "clientHeight", {
    value: 200
  });

  const target = dom.window.document.querySelector(".target");
  const { wrapper } = setup(mount, undefined, target);

  wrapper.instance().setComponentsCSSTransform();

  expect(setComponentsCSSTransform).toHaveLastReturnedWith({});

  wrapper.setProps({ selectedFonts });
  wrapper.setState({ stage: 4 });
  wrapper.instance().setComponentsCSSTransform();

  expect(setComponentsCSSTransform).toHaveLastReturnedWith({
    transform: "translateY(200px)"
  });

  wrapper.setState({ stage: 2 });
  wrapper.instance().setComponentsCSSTransform();

  expect(setComponentsCSSTransform).toHaveLastReturnedWith({
    transform: "translateY(-200px)"
  });

  wrapper.setState({ stage: 5 });
  wrapper.instance().setComponentsCSSTransform();

  expect(setComponentsCSSTransform).toHaveLastReturnedWith({
    transform: "translateY(200px)"
  });
});

it("should change component's stage when fonts added to the list", () => {
  const { wrapper } = setup();

  expect(wrapper.state("stage")).toBe(0);

  wrapper.setProps({ selectedFonts: ["Roboto"] });

  expect(wrapper.state("stage")).toBe(1);
});

it("should change component's stage when the list becomes empty", () => {
  const { wrapper } = setup();

  expect(wrapper.state("stage")).toBe(0);

  wrapper.setProps({ selectedFonts: ["Roboto"] });

  expect(wrapper.state("stage")).toBe(1);

  wrapper.setProps({ selectedFonts: [] });

  expect(wrapper.state("stage")).toBe(0);
});

it("should set corresponding component classes", () => {
  const { wrapper } = setup(mount);

  expect(wrapper.state("stage")).toBe(0);
  expect(wrapper.find(".SelectedFonts").hasClass("empty")).toBeTruthy();

  wrapper.setProps({ selectedFonts });
  wrapper.update();
  expect(wrapper.find(".SelectedFonts").hasClass("folded")).toBeTruthy();

  wrapper.setState({ stage: 2 });
  expect(wrapper.find(".SelectedFonts").hasClass("unfolding")).toBeTruthy();

  wrapper.setState({ stage: 3 });
  expect(wrapper.find(".SelectedFonts").hasClass("active")).toBeTruthy();

  wrapper.setState({ stage: 4 });
  expect(wrapper.find(".SelectedFonts").hasClass("folding")).toBeTruthy();

  wrapper.setState({ stage: 5 });
  expect(
    wrapper.find(".SelectedFonts").hasClass("folding-to-clear")
  ).toBeTruthy();
});

it("should set corresponding title text", () => {
  const { wrapper } = setup();

  expect(wrapper.find(".SelectedFonts").prop("title")).toBe(
    "Select fonts first"
  );

  wrapper.setProps({ selectedFonts });

  expect(wrapper.find(".SelectedFonts").prop("title")).toBe("Click to Expand");
});

it("should re-render on state update or selected fonts list length change", () => {
  const { wrapper } = setup(mount);
  const shouldComponentUpdate = jest.spyOn(
    SelectedFonts.prototype,
    "shouldComponentUpdate"
  );

  wrapper.setProps({ selectedFonts });

  // First call triggered by props update.
  // Second call triggered by state update (selectedFonts.length
  // change causes "stage" update)
  expect(shouldComponentUpdate).toHaveBeenCalledTimes(2);
  expect(shouldComponentUpdate).toHaveLastReturnedWith(true);

  wrapper.setState({ stage: 2 });

  expect(shouldComponentUpdate).toHaveBeenCalledTimes(3);
  expect(shouldComponentUpdate).toHaveLastReturnedWith(true);

  const updatedSelectedFonts = [...selectedFonts];
  wrapper.setProps({ selectedFonts: updatedSelectedFonts });

  expect(shouldComponentUpdate).toHaveBeenCalledTimes(4);
  expect(shouldComponentUpdate).toHaveLastReturnedWith(false);
});
