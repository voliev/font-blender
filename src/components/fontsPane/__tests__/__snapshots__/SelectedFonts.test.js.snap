// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not crash if the list is empty 1`] = `
<Fragment>
  <div
    className="SelectedFonts empty"
    onTransitionEnd={[Function]}
    style={Object {}}
    title="Select fonts first"
  >
    <OutsideClickHandler
      disabled={true}
      display="block"
      onOutsideClick={[Function]}
      useCapture={true}
    >
      <SelectedFontsHeaderBar
        folded={true}
        headerBarClickHandler={[Function]}
        selectedFonts={Array []}
      />
    </OutsideClickHandler>
  </div>
  <div
    className="overlay"
  />
</Fragment>
`;

exports[`should render without crashing 1`] = `
<Fragment>
  <div
    className="SelectedFonts empty"
    onTransitionEnd={[Function]}
    style={Object {}}
    title="Select fonts first"
  >
    <OutsideClickHandler
      disabled={true}
      display="block"
      onOutsideClick={[Function]}
      useCapture={true}
    >
      <SelectedFontsHeaderBar
        folded={true}
        headerBarClickHandler={[Function]}
        selectedFonts={Array []}
      />
    </OutsideClickHandler>
  </div>
  <div
    className="overlay"
  />
</Fragment>
`;
